heartbeat/osceip

```diff
@@ -0,0 +1,233 @@
+</parameter>
+
+<parameter name="elastic_ip" unique="1" required="1">
+<longdesc lang="en">
+reserved elastic ip for vm instance
+</longdesc>
+<shortdesc lang="en">reserved elastic ip for vm instance</shortdesc>
+<content type="string" default="" />
+</parameter>
+
+<parameter name="region" required="0">
+<longdesc lang="en">
+Region for osc resource (required for role-based authentication)
+</longdesc>
+<shortdesc lang="en">Region</shortdesc>
+<content type="string" default="${OCF_RESKEY_region_default}" />
+</parameter>
+</parameters>
+
+<actions>
+<action name="start"        timeout="30s" />
+<action name="stop"         timeout="30s" />
+<action name="monitor"      timeout="30s" interval="20s" depth="0" />
+<action name="migrate_to"   timeout="30s" />
+<action name="migrate_from" timeout="30s" />
+<action name="meta-data"    timeout="5s" />
+<action name="validate"     timeout="10s" />
+<action name="validate-all" timeout="10s" />
+</actions>
+</resource-agent>
+END
+}
+
+#######################################################################
+
+osceip_usage() {
+    cat <<END
+usage: $0 {start|stop|monitor|migrate_to|migrate_from|validate|validate-all|meta-data}
+
+Expects to have a fully populated OCF RA-compliant environment set.
+END
+}
+
+osceip_start() {
+    osceip_monitor || return $OCF_SUCCESS
+
+    $oscCLI_CMD api LinkPublicIp --profile "${OCF_RESKEY_profile}" --PublicIp "${ELASTIC_IP}" --VmId "${INSTANCE_ID}" --AllowRelink True
+    RET=$?
+
+
+    if [ $RET -ne 0 ]; then
+        return $OCF_NOT_RUNNING
+    fi
+
+    ocf_log info "elastic_ip has been successfully brought up (${ELASTIC_IP})"
+    return $OCF_SUCCESS
+}
+
+osceip_stop() {
+    osceip_monitor && return $OCF_SUCCESS
+
+    $oscCLI_CMD api UnlinkPublicIp --profile "${OCF_RESKEY_profile}" --PublicIp "${ELASTIC_IP}"
+    RET=$?
+
+    if [ $RET -ne 0 ]; then
+        return $OCF_NOT_RUNNING
+    fi
+
+    ocf_log info "elastic_ip has been successfully brought down (${ELASTIC_IP})"
+    return $OCF_SUCCESS
+}
+
+osceip_monitor() {
+    $oscCLI_CMD api ReadPublicIps --profile "${OCF_RESKEY_profile}" --Filters '{"VmIds": ["'${INSTANCE_ID}'"],}' |grep LinkPublicIpId | awk -F'"' '{print$4}'
+    RET=$?
+
+    if [ $RET -ne 0 ]; then
+        return $OCF_NOT_RUNNING
+    fi
+    return $OCF_SUCCESS
+}
+
+osceip_validate() {
+    check_binary "${OCF_RESKEY_osccli}"
+
+    if [ -z "${INSTANCE_ID}" ]; then
+        ocf_exit_reason "instance_id not found. Is this a vm instance?"
+        return $OCF_ERR_GENERIC
+    fi
+
+    return $OCF_SUCCESS
+}
+
+case $__OCF_ACTION in
+    meta-data)
+        meta_data
+        exit $OCF_SUCCESS
+        ;;
+    usage|help)
+        osceip_usage
+        exit $OCF_SUCCESS
+        ;;
+esac
+
+oscCLI_CMD="${OCF_RESKEY_osccli}"
+if [ -n "${OCF_RESKEY_region}" ]; then
+	oscCLI_CMD="$oscCLI_CMD --region ${OCF_RESKEY_region}"
+fi
+ELASTIC_IP="${OCF_RESKEY_elastic_ip}"
+INSTANCE_ID=$(curl -s http://169.254.169.254/latest/meta-data/instance-id)
```

Comment:
You should use curl_retry() here, and add the parameters for it, so it wont fail if there's a minor hickup in the network or API servers. See the aws-* agents for how to do that.

Reply:
I have a strange behavior with curl_retry

./osceip: 222: curl_retry: not found
INSTANCE_ID=$(curl_retry "${OCF_RESKEY_curl_retries}" "${OCF_RESKEY_curl_sleep}" -s" "http://169.254.169.254/latest/meta-data/instance-id")

even if  i have :
. ${OCF_FUNCTIONS_DIR}/ocf-shellfuncs

Reply:
Ah. It has only been added recently:
https://github.com/ClusterLabs/resource-agents/commit/fc0657b936f6a58f741e33f851b22f82bc68bffa

You can set `OCF_ROOT=. OCF_FUNCTIONS_DIR=.` when trying to run it manually (after running `make`).

Reply:
fixed